File: people/kaiemjoy/cv/fetch_from_ncbi.py
#!/usr/bin/env python3
"""
Fetch and format publications from NCBI for Dr. Kristen Aiemjoy's CV
Fetches from her personal NCBI My Bibliography profile
Formats in LaTeX CV style: Last name, First Middle initial. with hyperlinks
"""

import json
import urllib.request
import urllib.error
import urllib.parse
import sys
import xml.etree.ElementTree as ET
from datetime import datetime
import yaml
import re

def fetch_pubmed_ids(author="Aiemjoy K[Author]", max_results=200):
    """
    Fetch PubMed IDs for Dr. Aiemjoy's publications using E-utilities
    """
    search_url = f"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term={urllib.parse.quote(author)}&retmax={max_results}&retmode=json&sort=pub+date"
    
    try:
        with urllib.request.urlopen(search_url, timeout=30) as response:
            data = json.loads(response.read().decode())
            pmids = data.get('esearchresult', {}).get('idlist', [])
            print(f"Found {len(pmids)} publications from PubMed", file=sys.stderr)
            return pmids
    except Exception as e:
        print(f"Error fetching from PubMed: {e}", file=sys.stderr)
        return []

def fetch_pubmed_details_xml(pmids):
    """
    Fetch detailed publication information in XML format
    """
    if not pmids:
        return None
    
    # Batch process in chunks of 100
    all_xml = []
    for i in range(0, len(pmids), 100):
        batch = pmids[i:i+100]
        pmid_str = ",".join(batch)
        fetch_url = f"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&id={pmid_str}&retmode=xml"
        
        try:
            print(f"Fetching batch {i//100 + 1}...", file=sys.stderr)
            with urllib.request.urlopen(fetch_url, timeout=60) as response:
                xml_data = response.read().decode()
                all_xml.append(xml_data)
        except Exception as e:
            print(f"Error fetching publication details: {e}", file=sys.stderr)
    
    # Combine XML
    if all_xml:
        return '\n'.join(all_xml)
    return None

def format_author_name(last, first, initials=''):
    """
    Format author name as: Last F, or Last FM if middle initial exists
    Aiemjoy should be bolded
    """
    # Get first initial
    first_initial = first[0].upper() if first else ''
    
    # Get middle initial if present
    middle_initial = initials[0].upper() if initials and len(initials) > 0 else ''
    
    # Format: Last FI or Last F
    if middle_initial:
        formatted = f"{last} {first_initial}{middle_initial}"
    else:
        formatted = f"{last} {first_initial}"
    
    # Bold Aiemjoy
    if last == "Aiemjoy":
        formatted = f"**{formatted}**"
    
    return formatted

def parse_pubmed_xml(xml_data):
    """
    Parse PubMed XML and extract publication details
    """
    publications = []
    
    try:
        # Parse XML
        root = ET.fromstring(f'<root>{xml_data}</root>')
        
        for article in root.findall('.//PubmedArticle'):
            pub = {}
            
            # Get PMID
            pmid_elem = article.find('.//PMID')
            pub['pmid'] = pmid_elem.text if pmid_elem is not None else ''
            
            # Get title
            title_elem = article.find('.//ArticleTitle')
            if title_elem is not None:
                pub['title'] = ''.join(title_elem.itertext()).strip()
            else:
                pub['title'] = ''
            
            # Get authors
            authors = []
            author_list = article.find('.//AuthorList')
            if author_list is not None:
                for author in author_list.findall('Author'):
                    last_elem = author.find('LastName')
                    first_elem = author.find('ForeName')
                    initials_elem = author.find('Initials')
                    
                    if last_elem is not None:
                        last = last_elem.text
                        first = first_elem.text if first_elem is not None else ''
                        initials = initials_elem.text if initials_elem is not None else ''
                        
                        formatted_author = format_author_name(last, first, initials)
                        authors.append(formatted_author)
            
            pub['authors'] = authors
            
            # Get journal
            journal_elem = article.find('.//Journal/Title')
            if journal_elem is None:
                journal_elem = article.find('.//Journal/ISOAbbreviation')
            pub['journal'] = journal_elem.text if journal_elem is not None else ''
            
            # Get publication date
            pub_date = article.find('.//PubDate')
            year = ''
            month = ''
            day = ''
            if pub_date is not None:
                year_elem = pub_date.find('Year')
                month_elem = pub_date.find('Month')
                day_elem = pub_date.find('Day')
                year = year_elem.text if year_elem is not None else ''
                month = month_elem.text if month_elem is not None else ''
                day = day_elem.text if day_elem is not None else ''
            
            pub['year'] = year
            pub['month'] = month
            pub['day'] = day
            
            # Get volume, issue, pages
            volume_elem = article.find('.//Volume')
            issue_elem = article.find('.//Issue')
            pages_elem = article.find('.//MedlinePgn')
            
            pub['volume'] = volume_elem.text if volume_elem is not None else ''
            pub['issue'] = issue_elem.text if issue_elem is not None else ''
            pub['pages'] = pages_elem.text if pages_elem is not None else ''
            
            # Get DOI
            doi = ''
            for article_id in article.findall('.//ArticleId'):
                if article_id.get('IdType') == 'doi':
                    doi = article_id.text
                    break
            pub['doi'] = doi
            
            publications.append(pub)
            
    except Exception as e:
        print(f"Error parsing XML: {e}", file=sys.stderr)
    
    return publications

def format_publication_for_cv(pub):
    """
    Format publication in the requested LaTeX CV style:
    Authors. [hyperlinked title]. Journal. Date;volume(issue):pages. PMID: xxxxx.
    """
    # Format authors
    author_str = ', '.join(pub['authors']) if pub['authors'] else ''
    
    # Format date
    date_parts = []
    if pub['year']:
        date_parts.append(pub['year'])
    if pub['month']:
        # Convert month name to abbreviated form if needed
        month_abbrev = pub['month'][:3] if len(pub['month']) > 3 else pub['month']
        date_parts.append(month_abbrev)
    if pub['day']:
        date_parts.append(pub['day'])
    date_str = ' '.join(date_parts)
    
    # Format volume/issue/pages
    citation_parts = []
    if pub['volume']:
        vol_str = pub['volume']
        if pub['issue']:
            vol_str += f"({pub['issue']})"
        citation_parts.append(vol_str)
    if pub['pages']:
        citation_parts.append(pub['pages'])
    citation_str = ':'.join(citation_parts) if citation_parts else ''
    
    # Create the formatted publication
    formatted = f"{author_str}. "
    
    # Add hyperlinked title
    if pub['doi']:
        formatted += f"[{pub['title']}](https://doi.org/{pub['doi']}). "
    else:
        formatted += f"{pub['title']}. "
    
    # Add journal in italics
    if pub['journal']:
        formatted += f"*{pub['journal']}*. "
    
    # Add date and citation
    if date_str and citation_str:
        formatted += f"{date_str};{citation_str}. "
    elif date_str:
        formatted += f"{date_str}. "
    
    # Add PMID
    if pub['pmid']:
        formatted += f"PMID: {pub['pmid']}."
    
    return formatted.strip()

def create_publications_yaml(publications):
    """
    Create YAML format for publications with formatted strings
    """
    yaml_pubs = []
    
    for pub in publications:
        yaml_pub = {
            'formatted': format_publication_for_cv(pub),
            'year': pub['year'],
            'pmid': pub['pmid'],
            'doi': pub['doi'],
            'title': pub['title']
        }
        yaml_pubs.append(yaml_pub)
    
    return yaml_pubs

def main():
    print("Fetching publications from Dr. Aiemjoy's NCBI profile...", file=sys.stderr)
    
    # Fetch PubMed IDs
    pmids = fetch_pubmed_ids("Aiemjoy K[Author]")
    
    if not pmids:
        print("No publications found", file=sys.stderr)
        sys.exit(1)
    
    # Fetch details
    xml_data = fetch_pubmed_details_xml(pmids)
    
    if not xml_data:
        print("Failed to fetch publication details", file=sys.stderr)
        sys.exit(1)
    
    # Parse publications
    publications = parse_pubmed_xml(xml_data)
    
    print(f"Parsed {len(publications)} publications", file=sys.stderr)
    
    # Create YAML
    yaml_pubs = create_publications_yaml(publications)
    
    # Sort by year (descending)
    yaml_pubs.sort(key=lambda x: x['year'], reverse=True)
    
    # Write to file
    output_file = 'people/kaiemjoy/cv/publications.yml'
    with open(output_file, 'w') as f:
        yaml.dump(yaml_pubs, f, default_flow_style=False, sort_keys=False, allow_unicode=True)
    
    print(f"Saved {len(yaml_pubs)} publications to {output_file}", file=sys.stderr)
    
    # Also print first few for verification
    print("\nFirst 3 publications:", file=sys.stderr)
    for i, pub in enumerate(yaml_pubs[:3]):
        print(f"\n{i+1}. {pub['formatted']}", file=sys.stderr)

if __name__ == "__main__":
    main()

File: people/kaiemjoy/cv/fetch_from_ncbi.R
#!/usr/bin/env Rscript
#
# fetch_from_ncbi.R
#
# Fetch publications from NCBI My Bibliography for Dr. Kristen Aiemjoy
# using NCBI E-utilities API
#
# Usage: Rscript fetch_from_ncbi.R
#
# Output: publications.yml in LaTeX CV format
#
# NCBI My Bibliography Collection:
# https://www.ncbi.nlm.nih.gov/myncbi/1xIGpkekG9FQP/bibliography/public/

# Check and load required packages
required_packages <- c("yaml", "jsonlite", "xml2")
for (pkg in required_packages) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    message("Installing required package: ", pkg)
    install.packages(pkg, repos = "https://cloud.r-project.org")
  }
  suppressPackageStartupMessages(library(pkg, character.only = TRUE))
}

# NCBI E-utilities base URLs
ESEARCH_URL <- "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi"
EFETCH_URL <- "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi"
ESUMMARY_URL <- "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi"

#' Fetch PubMed IDs for an author
#'
#' @param author_name Author name (e.g., "Aiemjoy K")
#' @param retmax Maximum number of results to return
#' @return Vector of PubMed IDs
fetch_pubmed_ids <- function(author_name, retmax = 100) {
  # Build search query
  query <- paste0(author_name, "[Author]")
  
  # Construct URL
  url <- paste0(
    ESEARCH_URL,
    "?db=pubmed",
    "&term=", URLencode(query, reserved = TRUE),
    "&retmax=", retmax,
    "&retmode=json"
  )
  
  cat("Fetching PubMed IDs for:", author_name, "\n")
  cat("URL:", url, "\n")
  
  # Fetch data
  tryCatch({
    response <- readLines(url, warn = FALSE)
    data <- jsonlite::fromJSON(paste(response, collapse = ""))
    
    if (!is.null(data$esearchresult$idlist)) {
      pmids <- data$esearchresult$idlist
      cat("Found", length(pmids), "publications\n")
      pmids
    } else {
      cat("No publications found\n")
      character(0)
    }
  }, error = function(e) {
    cat("Error fetching PubMed IDs:", e$message, "\n")
    character(0)
  })
}

#' Fetch publication details for PubMed IDs
#'
#' @param pmids Vector of PubMed IDs
#' @return List of publication records
fetch_publication_details <- function(pmids) {
  if (length(pmids) == 0) {
    cat("No PMIDs provided\n")
    return(list())
  }
  
  # Process in batches of 100
  batch_size <- 100
  all_pubs <- list()
  
  for (i in seq(1, length(pmids), by = batch_size)) {
    batch_end <- min(i + batch_size - 1, length(pmids))
    batch_pmids <- pmids[i:batch_end]
    
    cat("Fetching details for PMIDs", i, "to", batch_end, "\n")
    
    # Construct URL for efetch
    pmid_str <- paste(batch_pmids, collapse = ",")
    url <- paste0(
      EFETCH_URL,
      "?db=pubmed",
      "&id=", pmid_str,
      "&retmode=xml"
    )
    
    tryCatch({
      # Fetch XML data
      xml_data <- xml2::read_xml(url)
      
      # Parse each article
      articles <- xml2::xml_find_all(xml_data, ".//PubmedArticle")
      
      for (article in articles) {
        pub <- parse_pubmed_article(article)
        if (!is.null(pub)) {
          all_pubs <- c(all_pubs, list(pub))
        }
      }
      
      # Be polite to NCBI servers
      Sys.sleep(0.4)
    }, error = function(e) {
      cat("Error fetching details:", e$message, "\n")
    })
  }
  
  cat("Successfully parsed", length(all_pubs), "publications\n")
  all_pubs
}

#' Parse a PubMed article XML node
#'
#' @param article XML node for a PubmedArticle
#' @return List with publication details
parse_pubmed_article <- function(article) {
  tryCatch({
    # Extract basic info
    medline <- xml2::xml_find_first(article, ".//MedlineCitation")
    pmid <- xml2::xml_text(
      xml2::xml_find_first(medline, ".//PMID")
    )
    
    # Extract article details
    art <- xml2::xml_find_first(medline, ".//Article")
    
    # Title
    title <- xml2::xml_text(
      xml2::xml_find_first(art, ".//ArticleTitle")
    )
    
    # Journal
    journal <- xml2::xml_text(
      xml2::xml_find_first(art, ".//Journal/Title")
    )
    
    # Authors
    author_nodes <- xml2::xml_find_all(art, ".//Author")
    authors <- sapply(author_nodes, function(auth) {
      last_name <- xml2::xml_text(
        xml2::xml_find_first(auth, ".//LastName")
      )
      initials <- xml2::xml_text(
        xml2::xml_find_first(auth, ".//Initials")
      )
      if (nchar(last_name) > 0 && nchar(initials) > 0) {
        paste(last_name, initials)
      } else {
        NA
      }
    })
    authors <- authors[!is.na(authors)]
    
    # Publication date
    pub_date_node <- xml2::xml_find_first(art, ".//PubDate")
    year <- xml2::xml_text(
      xml2::xml_find_first(pub_date_node, ".//Year")
    )
    month <- xml2::xml_text(
      xml2::xml_find_first(pub_date_node, ".//Month")
    )
    day <- xml2::xml_text(
      xml2::xml_find_first(pub_date_node, ".//Day")
    )
    
    # Format date
    if (nchar(month) > 0) {
      month_num <- match(
        month,
        c("Jan", "Feb", "Mar", "Apr", "May", "Jun",
          "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
      )
      if (is.na(month_num)) {
        month_num <- as.integer(month)
      }
      if (nchar(day) > 0) {
        date <- sprintf("%s-%02d-%02d",
                       year, month_num, as.integer(day))
      } else {
        date <- sprintf("%s-%02d-01", year, month_num)
      }
    } else {
      date <- paste0(year, "-01-01")
    }
    
    # DOI
    doi_node <- xml2::xml_find_first(
      art,
      ".//ELocationID[@EIdType='doi']"
    )
    if (length(doi_node) > 0) {
      doi <- xml2::xml_text(doi_node)
    } else {
      doi <- NA
    }
    
    # Return publication record
    list(
      id = paste0("pmid-", pmid),
      title = title,
      authors = authors,
      date = date,
      journal = journal,
      doi = if (!is.na(doi)) doi else NULL,
      pmid = pmid
    )
  }, error = function(e) {
    cat("Error parsing article:", e$message, "\n")
    NULL
  })
}

#' Format authors in LaTeX CV style with bolded Aiemjoy
#'
#' @param authors Vector of author names in "Last FI" format
#' @return Formatted author string with HTML tags
format_authors_latex <- function(authors) {
  formatted <- sapply(authors, function(auth) {
    if (grepl("Aiemjoy", auth, ignore.case = TRUE)) {
      paste0("<strong>", auth, "</strong>")
    } else {
      auth
    }
  })
  paste(formatted, collapse = ", ")
}

#' Save publications to YAML file in LaTeX CV format
#'
#' @param publications List of publication records
#' @param output_file Output YAML file path
save_publications_yaml <- function(publications, output_file) {
  # Format publications for YAML
  yaml_pubs <- lapply(publications, function(pub) {
    # Format authors
    author_str <- format_authors_latex(pub$authors)
    
    # Create title with hyperlink if DOI exists
    if (!is.null(pub$doi) && !is.na(pub$doi)) {
      doi_url <- paste0("https://doi.org/", pub$doi)
      title_html <- paste0(
        "<a href=\"", doi_url, "\">", pub$title, "</a>"
      )
    } else {
      title_html <- pub$title
    }
    
    # Format journal in italics
    journal_html <- paste0("<em>", pub$journal, "</em>")
    
    # Extract year
    year <- substr(pub$date, 1, 4)
    
    # Build description in LaTeX CV format
    desc_parts <- c(
      author_str,
      title_html,
      journal_html,
      year
    )
    
    if (!is.null(pub$doi) && !is.na(pub$doi)) {
      desc_parts <- c(desc_parts, paste0("DOI: ", pub$doi))
    }
    
    description <- paste(desc_parts, collapse = ". ")
    description <- paste0(description, ".")
    
    list(
      id = pub$id,
      title = pub$title,
      authors = pub$authors,
      date = pub$date,
      journal = pub$journal,
      doi = pub$doi,
      pmid = pub$pmid,
      description = description
    )
  })
  
  # Write to YAML
  yaml::write_yaml(yaml_pubs, output_file)
  cat("Saved", length(yaml_pubs), "publications to", output_file, "\n")
}

#' Main function
main <- function() {
  cat("=== Fetching Dr. Aiemjoy's publications from NCBI ===\n\n")
  
  # Fetch PMIDs
  pmids <- fetch_pubmed_ids("Aiemjoy K", retmax = 100)
  
  if (length(pmids) == 0) {
    cat("No publications found. Exiting.\n")
    return(invisible(NULL))
  }
  
  # Fetch details
  publications <- fetch_publication_details(pmids)
  
  if (length(publications) == 0) {
    cat("No publication details retrieved. Exiting.\n")
    return(invisible(NULL))
  }
  
  # Sort by date (most recent first)
  publications <- publications[
    order(
      sapply(publications, function(p) p$date),
      decreasing = TRUE
    )
  ]
  
  # Determine output file path
  # Try to get script location, fall back to current directory
  script_path <- tryCatch({
    normalizePath(sys.frame(1)$ofile)
  }, error = function(e) {
    # If that fails, try commandArgs
    tryCatch({
      args <- commandArgs(trailingOnly = FALSE)
      file_arg <- grep("^--file=", args, value = TRUE)
      if (length(file_arg) > 0) {
        normalizePath(sub("^--file=", "", file_arg))
      } else {
        NULL
      }
    }, error = function(e2) NULL)
  })
  
  if (!is.null(script_path) && file.exists(script_path)) {
    output_file <- file.path(dirname(script_path), "publications.yml")
  } else {
    # Fall back to current working directory
    output_file <- "publications.yml"
  }
  
  save_publications_yaml(publications, output_file)
  
  cat("\n=== Complete! ===\n")
  cat("Publications saved to:", output_file, "\n")
  cat("Total publications:", length(publications), "\n")
  
  invisible(publications)
}

# Run main function if script is executed directly
if (!interactive()) {
  main()
}
